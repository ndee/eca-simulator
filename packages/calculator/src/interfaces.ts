export enum PurchaseRepeat {
  Daily = "Daily",
  Weekly = "Weekly",
  Monthly = "Monthly",
  Yearly = "Yearly",
}

export interface IEcaPurchase {
  purchaseAmountInEur: number;
  purchaseRepeat: PurchaseRepeat;
  purchaseStart: Date;
  purchaseEnd: Date;
}
export interface IEcaResult {
  acquiredSatoshis: number;
  investedEur: number;
}

export interface IPurchaseRequest {
  amountInEur: number;
  date: Date;
}

export interface IPurchase extends IPurchaseRequest {
  btcUsdRate: number;
  usdEurRate: number;
  acquiredSatoshis: number;
  exchangeRateSource?: string;
}

export enum ExchangeRateType {
  BtcToUsd = "BtcToUsd",
  UsdToEur = "UsdToEur",
}

export type ExchangeRateProvider = (
  dates: Date[],
  rateType: ExchangeRateType
) => Promise<number[]>;

export interface IBuildPurchaseParams {
  requestList: IPurchaseRequest[];
  exchangeRateProvider: ExchangeRateProvider;
}
