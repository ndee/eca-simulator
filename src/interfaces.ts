export enum PurchaseRepeat {
  Daily = "Daily",
  Weekly = "Weekly",
  Monthly = "Monthly",
  Yearly = "Yearly",
}

export interface IEcaPurchase {
  purchaseAmountInEur: number;
  purchaseRepeat: PurchaseRepeat;
  purchaseStart: Date;
  purchaseEnd: Date;
}
export interface IEcaResult {
  acquiredSatoshis: number;
  investedEur: number;
}

export interface IPurchaseRequest {
  amountInEur: number;
  date: Date;
}
