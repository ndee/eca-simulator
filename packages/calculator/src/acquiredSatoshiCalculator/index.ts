import { IEcaPurchase, IEcaResult } from "../interfaces";

export async function calculateAcquiredBitcoin(
  ecaPurchase: IEcaPurchase
): Promise<IEcaResult> {
  return { acquiredSatoshis: 1234, investedEur: 12 } as IEcaResult;
}
