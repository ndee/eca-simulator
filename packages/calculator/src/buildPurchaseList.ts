import { calculateSatsFromEur } from "./calculateSatsFromEur";
import {
  ExchangeRateType,
  IBuildPurchaseParams,
  IPurchase,
} from "./interfaces";

export async function buildPurchaseList(
  params: IBuildPurchaseParams
): Promise<IPurchase[]> {
  const btcUsdRates = await params.exchangeRateProvider(
    params.requestList.map((r) => r.date),
    ExchangeRateType.BtcToUsd
  );
  const usdEurRates = await params.exchangeRateProvider(
    params.requestList.map((r) => r.date),
    ExchangeRateType.UsdToEur
  );

  const purchases: IPurchase[] = params.requestList.map((request, index) => ({
    date: request.date,
    amountInEur: request.amountInEur,
    btcUsdRate: btcUsdRates[index],
    usdEurRate: usdEurRates[index],
    acquiredSatoshis: calculateSatsFromEur({
      amountInEur: request.amountInEur,
      btcUsdRate: btcUsdRates[index],
      usdEurRate: usdEurRates[index],
    }),
  }));

  return purchases;
}
