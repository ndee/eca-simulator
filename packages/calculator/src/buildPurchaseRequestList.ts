import moment from "moment";
import { IEcaPurchase, IPurchaseRequest, PurchaseRepeat } from "./interfaces";

export function buildPurchaseRequestList(
  ecaPurchase: IEcaPurchase
): IPurchaseRequest[] {
  if (ecaPurchase.purchaseStart > ecaPurchase.purchaseEnd) {
    throw new Error(
      "Purchase start date cannot be later than purchase end date"
    );
  }
  const requests: IPurchaseRequest[] = [];
  let purchaseMoment = moment(ecaPurchase.purchaseStart);

  while (purchaseMoment.isBefore(moment(ecaPurchase.purchaseEnd))) {
    let momentUnit: moment.unitOfTime.DurationConstructor;
    switch (ecaPurchase.purchaseRepeat) {
      case PurchaseRepeat.Yearly:
        momentUnit = "year";
        break;
      case PurchaseRepeat.Monthly:
        momentUnit = "month";
        break;
      case PurchaseRepeat.Weekly:
        momentUnit = "week";
        break;
      case PurchaseRepeat.Daily:
        momentUnit = "day";
        break;
    }
    purchaseMoment.add(1, momentUnit);
    requests.push({
      amountInEur: ecaPurchase.purchaseAmountInEur,
      date: purchaseMoment.toDate(),
    });
  }

  return requests;
}
