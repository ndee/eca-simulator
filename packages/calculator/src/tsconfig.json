{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "outDir": "../dist",
    "composite": true,
    "declaration": true,
    "skipLibCheck": true
  },
  "include": ["**/*", "../test/exchangeRateProvider/getBtcRateFromTimeChainCalendar.spec.ts"]
}
