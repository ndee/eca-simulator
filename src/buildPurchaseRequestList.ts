import moment from "moment";
import { IEcaPurchase, IPurchaseRequest, PurchaseRepeat } from "./interfaces";

export function buildPurchaseRequestList(
  ecaPurchase: IEcaPurchase
): IPurchaseRequest[] {
  if (ecaPurchase.purchaseStart > ecaPurchase.purchaseEnd) {
    throw new Error(
      "Purchase start date cannot be later than purchase end date"
    );
  }
  const requests: IPurchaseRequest[] = [];
  let purchaseMoment = moment(ecaPurchase.purchaseStart);

  while (purchaseMoment.date() < ecaPurchase.purchaseEnd.getDate()) {
    switch (ecaPurchase.purchaseRepeat) {
      case PurchaseRepeat.Yearly:
        purchaseMoment.add(1, "year");
        break;
      case PurchaseRepeat.Monthly:
        purchaseMoment.add(1, "month");
        break;
      case PurchaseRepeat.Weekly:
        purchaseMoment.add(1, "week");
        break;
      case PurchaseRepeat.Daily:
        purchaseMoment.add(1, "day");
        break;
    }
    requests.push({
      amountInEur: ecaPurchase.purchaseAmountInEur,
      date: purchaseMoment.toDate(),
    });
  }

  return requests;
}
