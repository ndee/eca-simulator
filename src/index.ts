import { getBitcoinPrice } from "./getBitcoinPrice";

const day = new Date(2021, 9, 3);
getBitcoinPrice(day).then((price) => {
  console.log(
    `The price of Bitcoin at ${day.toDateString()} was: ${price} USD`
  );
});
