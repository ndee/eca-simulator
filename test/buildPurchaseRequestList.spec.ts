import { buildPurchaseRequestList } from "../src/buildPurchaseRequestList";
import { PurchaseRepeat } from "../src/interfaces";

describe("buildPurchaseRequestList", () => {
  describe("should calculate list for", () => {
    it("weekly plan", () => {
      expect(
        buildPurchaseRequestList({
          purchaseAmountInEur: 10,
          purchaseStart: new Date("2022-01-01"),
          purchaseEnd: new Date("2022-05-26"),
          purchaseRepeat: PurchaseRepeat.Weekly,
        })
      ).toMatchSnapshot();
    });
  });
});
