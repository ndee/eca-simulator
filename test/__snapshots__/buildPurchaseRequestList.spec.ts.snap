// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildPurchaseRequestList should calculate list for a daily plan 1`] = `
[
  {
    "amountInEur": 1,
    "date": 2022-03-02T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-03T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-04T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-05T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-06T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-07T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-08T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-09T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-10T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-11T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-12T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-13T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-14T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-15T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-16T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-17T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-18T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-19T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-20T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-21T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-22T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-23T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-24T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-25T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-26T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-27T00:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-27T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-28T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-29T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-30T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-03-31T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-01T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-02T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-03T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-04T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-05T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-06T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-07T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-08T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-09T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-10T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-11T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-12T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-13T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-14T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-15T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-16T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-17T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-18T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-19T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-20T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-21T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-22T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-23T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-24T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-25T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-26T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-27T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-28T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-29T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-04-30T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-01T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-02T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-03T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-04T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-05T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-06T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-07T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-08T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-09T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-10T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-11T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-12T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-13T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-14T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-15T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-16T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-17T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-18T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-19T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-20T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-21T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-22T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-23T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-24T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-25T23:00:00.000Z,
  },
  {
    "amountInEur": 1,
    "date": 2022-05-26T23:00:00.000Z,
  },
]
`;

exports[`buildPurchaseRequestList should calculate list for a monthly plan 1`] = `
[
  {
    "amountInEur": 100,
    "date": 2021-02-01T00:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2021-03-01T00:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2021-03-31T23:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2021-04-30T23:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2021-05-31T23:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2021-06-30T23:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2021-07-31T23:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2021-08-31T23:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2021-09-30T23:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2021-11-01T00:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2021-12-01T00:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2022-01-01T00:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2022-02-01T00:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2022-03-01T00:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2022-03-31T23:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2022-04-30T23:00:00.000Z,
  },
  {
    "amountInEur": 100,
    "date": 2022-05-31T23:00:00.000Z,
  },
]
`;

exports[`buildPurchaseRequestList should calculate list for a weekly plan 1`] = `
[
  {
    "amountInEur": 10,
    "date": 2022-01-08T00:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-01-15T00:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-01-22T00:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-01-29T00:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-02-05T00:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-02-12T00:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-02-19T00:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-02-26T00:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-03-05T00:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-03-12T00:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-03-19T00:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-03-26T00:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-04-01T23:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-04-08T23:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-04-15T23:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-04-22T23:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-04-29T23:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-05-06T23:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-05-13T23:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-05-20T23:00:00.000Z,
  },
  {
    "amountInEur": 10,
    "date": 2022-05-27T23:00:00.000Z,
  },
]
`;

exports[`buildPurchaseRequestList should calculate list for a yearly plan 1`] = `
[
  {
    "amountInEur": 1000,
    "date": 2011-01-01T00:00:00.000Z,
  },
  {
    "amountInEur": 1000,
    "date": 2012-01-01T00:00:00.000Z,
  },
  {
    "amountInEur": 1000,
    "date": 2013-01-01T00:00:00.000Z,
  },
  {
    "amountInEur": 1000,
    "date": 2014-01-01T00:00:00.000Z,
  },
  {
    "amountInEur": 1000,
    "date": 2015-01-01T00:00:00.000Z,
  },
  {
    "amountInEur": 1000,
    "date": 2016-01-01T00:00:00.000Z,
  },
  {
    "amountInEur": 1000,
    "date": 2017-01-01T00:00:00.000Z,
  },
  {
    "amountInEur": 1000,
    "date": 2018-01-01T00:00:00.000Z,
  },
  {
    "amountInEur": 1000,
    "date": 2019-01-01T00:00:00.000Z,
  },
  {
    "amountInEur": 1000,
    "date": 2020-01-01T00:00:00.000Z,
  },
  {
    "amountInEur": 1000,
    "date": 2021-01-01T00:00:00.000Z,
  },
  {
    "amountInEur": 1000,
    "date": 2022-01-01T00:00:00.000Z,
  },
  {
    "amountInEur": 1000,
    "date": 2023-01-01T00:00:00.000Z,
  },
]
`;
